vignettes/python-primer.Rmd

```diff
@@ -0,0 +1,920 @@
+class MyClass:
+  def a_method(self):
+    print("self.a_method() was called with", self)
+
+inst = MyClass()
+inst.a_method()
+
+MyClass.a_method()     # error, missing required argument `self`
+MyClass.a_method(inst) # identical to inst.a_method()
+```
+
+Other dunder's worth knowing about are:
+
+-   `__getitem__`: the function invoked when subsetting an instance with
+    `[` (Equivalent to defining a `[` S3 method in R.
+
+-   `__getattr__`: the function invoked when subsetting with `.`
+    (Equivalent to defining a `$` S3 method in R.
+
+-   `__iter__` and `__next__`: functions invoked by `for`.
+
+-   `__call__`: invoked when a class instance is called like a function
+    (e.g., `instance()`).
+
+-   `__bool__`: invoked by `if` and `while` (equivalent to
+    `as.logical()` in R, but returning only a scalar, not a vector).
+
+-   `__repr__`, `__str__`, functions invoked for formatting and pretty
+    printing (equivalent to `format()` and `print()` S3 generics in R.
+
+-   `__enter__` and `__exit__`: functions invoked by `with`.
+
+-   Many builtin Python functions are just sugar for invoking the
+    dunder. For example: calling `repr(x)` is identical to
+    `x.__repr__()`. Other builtins that are just sugar for invoking the
+    dunder are `next()`, `iter()`, `str()`, `list()`, `dict()`,
+    `bool()`, `dir()` and more!
+
+#### Iterators, revisited
+
+Now that we have the basics of `class`, it's time to revisit iterators.
+First, some terminology:
+
+**iterable**: something that can be iterated over. Concretely, a class
+that defines an `__iter__` method, whose job is to return an *iterator*.
+
+**iterator**: something that iterates. Concretely, a class that defines
+a `__next__` method, whose job is to return the next element each time
+it is called, and then raises a `StopIteration` exception once it's
+exhausted.
+
+It's common to see classes that are both iterables and iterators, where
+the `__iter__` method is just a stub that returns `self`.
+
+Here is a custom iterable / iterator implementation of Python's `range`
+(similar to `seq` in R)
+
+```{python, error = TRUE}
+class MyRange:
+  def __init__(self, start, end):
+    self.start = start
+    self.end = end
+
+  def __iter__(self):
+    # reset our counter.
+    self._index = self.start - 1
+    return self
+
+  def __next__(self):
+    if self._index < self.end:
+      self._index += 1 # increment
+      return self._index
+    else:
+      raise StopIteration
+
+for x in MyRange(1, 3):
+  print(x)
+
+# doing what `for` does, but manually
+r = MyRange(1, 3)
+it = iter(r)
+next(it)
+next(it)
+next(it)
+next(it)
+```
+
+### Decorators
+
+Decorators are just functions that take a function as an argument and
+then return a function. Any function can be invoked as a decorator with
+the `@` syntax, which is just sugar for this simple action:
+
+```{python}
+def my_decorator(func):
+  func.x = "a decorator modified this function by adding an attribute `x`"
```

Comment:
could this be made even simpler? Perhaps by removing the added attribute? So that it's even more clear that it's just a function call, basically...?

Reply:
Are you suggesting an identity function? I was looking for some minimal, simple-to-understand thing that demonstrates that decorators typically do *something* ... assigning an attribute seemed like the most accessible operation that came to mind. I don't want to go full bore and start talking about closures or all the machinery in `functools`...

Reply:
yeah... maybe i was thinking of something as simple as calling` tf_function()` ;-))) But you're right, adding an attriute makes sense.